# Exercise 1 - Iris dataset

1. Load the previously downloaded Iris dataset as a list of lists.

In [1]:
import csv
import math

with open("iris.csv") as f:
    ds = list(csv.reader(f))
    ds = ds[0: -1] # last line is empty so we remove it

2. Compute and print the mean and the standard deviation for each of the 4 measurement columns.

In [2]:
def mean_std(lst):
    m = sum(lst)/len(lst)
    std = math.sqrt(sum([(el - m) ** 2 for el in lst]) / len(lst))
    return m, std

In [3]:
for i in range(0, 4):
    lst = [line[i] for line in ds]
    lst = list(map(float, lst))
    m , std =  mean_std(lst)
    print("measurement", i, "has mean", m, " and standard deviation", std)

measurement 0 has mean 5.843333333333335  and standard deviation 0.8253012917851409
measurement 1 has mean 3.0540000000000007  and standard deviation 0.4321465800705435
measurement 2 has mean 3.7586666666666693  and standard deviation 1.7585291834055201
measurement 3 has mean 1.1986666666666672  and standard deviation 0.760612618588172


3. Compute and print the mean and the standard deviation for each of the 4 measurement columns, separately for each of the three Iris species.

In [4]:
lst = [line[4] for line in ds]
iris_type = set(lst)

In [5]:
for t in iris_type:
    for i in range(0, 4):
        lst = [line[i] for line in ds if line[4] == t]
        lst = list(map(float, lst))
        m , std =  mean_std(lst)
        print("for species", t, "the measurement", i, "has mean", m, " and standard deviation", std)

for species Iris-setosa the measurement 0 has mean 5.005999999999999  and standard deviation 0.348946987377739
for species Iris-setosa the measurement 1 has mean 3.4180000000000006  and standard deviation 0.37719490982779713
for species Iris-setosa the measurement 2 has mean 1.464  and standard deviation 0.17176728442867115
for species Iris-setosa the measurement 3 has mean 0.2439999999999999  and standard deviation 0.10613199329137278
for species Iris-virginica the measurement 0 has mean 6.587999999999998  and standard deviation 0.6294886813914925
for species Iris-virginica the measurement 1 has mean 2.9739999999999998  and standard deviation 0.319255383666431
for species Iris-virginica the measurement 2 has mean 5.552  and standard deviation 0.5463478745268441
for species Iris-virginica the measurement 3 has mean 2.026  and standard deviation 0.2718896835115301
for species Iris-versicolor the measurement 0 has mean 5.936  and standard deviation 0.5109833656783752
for species Iris-ver

4. Based on the results of exercises 2 and 3, which of the 4 measurements would you considering as being the most characterizing one for the three species?

According to excercise 2 `mesurement 2` has the biggest standard deviation.<br>
Moreover, according to exercise 3, the mean of `mesurement 2` taken separately for the 3 species are far from each other.<br>
So, `mesurement 2` is the most characterizing measurement.

5. Based on the considerations of Exercise 3, assign the flowers with the following measurements to what you consider would be the most likely species.

`5.2, 3.1, 4.0, 1.2`<br>
`4.9, 2.5, 5.6, 2.0`<br>
`5.4, 3.2, 1.9, 0.4`

The categories that seems to fit best are respectively:

`Iris-versicolor`<br>
`Iris-virginica`<br>
`Iris-setosa`